core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -335,18 +329,32 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
       (topicPartitions ++ addedTopicPartitions).toSet, newTxnStartTimestamp, updateTimestamp)
   }
 
-  def prepareAbortOrCommit(newState: TransactionState, updateTimestamp: Long): TxnTransitMetadata = {
-    prepareTransitionTo(newState, producerId, producerEpoch, lastProducerEpoch, txnTimeoutMs, topicPartitions.toSet,
-      txnStartTimestamp, updateTimestamp)
+  def prepareAbortOrCommit(newState: TransactionState, clientTransactionVersion: Short, nextProducerId: Long, updateTimestamp: Long): TxnTransitMetadata = {
+
+    val (updatedProducerEpoch, updatedLastProducerEpoch) = if (clientTransactionVersion >= 2) {
+      // We already ensured that we do not overflow here. MAX_SHORT is the highest possible value.
+      ((producerEpoch + 1).toShort, producerEpoch)
+    } else {
+      (producerEpoch, lastProducerEpoch)
+    }
+
+    prepareTransitionTo(newState, producerId, nextProducerId, updatedProducerEpoch, updatedLastProducerEpoch, txnTimeoutMs, topicPartitions.toSet,
+      txnStartTimestamp, updateTimestamp, clientTransactionVersion)
   }
 
   def prepareComplete(updateTimestamp: Long): TxnTransitMetadata = {
     val newState = if (state == PrepareCommit) CompleteCommit else CompleteAbort
 
     // Since the state change was successfully written to the log, unset the flag for a failed epoch fence
     hasFailedEpochFence = false
-    prepareTransitionTo(newState, producerId, producerEpoch, lastProducerEpoch, txnTimeoutMs, Set.empty[TopicPartition],
-      txnStartTimestamp, updateTimestamp)
+    val (updatedProducerId, updatedProducerEpoch) = if (clientTransactionVersion >= 2 && nextProducerId != RecordBatch.NO_PRODUCER_ID) {
```

Comment:
Echoing on one of my previous comment, it would be great to have `clientTransactionVersion >= 2` centralized in one place.